server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -50,187 +51,269 @@
+  )
+  {
+    ServerHolder topCandidate = null;
+    for (final Object2IntMap.Entry<String> entry : targetReplicants.object2IntEntrySet()) {
+      final int targetReplicantsInTier = entry.getIntValue();
+      if (targetReplicantsInTier <= 0) {
         continue;
       }
 
-      final int maxSegmentsInNodeLoadingQueue = params.getCoordinatorDynamicConfig()
-                                                      .getMaxSegmentsInNodeLoadingQueue();
+      final String tier = entry.getKey();
 
-      Predicate<ServerHolder> serverHolderPredicate;
-      if (maxSegmentsInNodeLoadingQueue > 0) {
-        serverHolderPredicate = s -> (s != null && s.getNumberOfSegmentsInQueue() < maxSegmentsInNodeLoadingQueue);
-      } else {
-        serverHolderPredicate = Objects::nonNull;
+      final List<ServerHolder> holders = getFilteredHolders(
+          tier,
+          params.getDruidCluster(),
+          createLoadQueueSizeLimitingPredicate(params)
+      );
+      if (holders.isEmpty()) {
+        continue;
       }
 
-      final List<ServerHolder> serverHolderList = serverQueue.stream()
-                                                             .filter(serverHolderPredicate)
-                                                             .collect(Collectors.toList());
-
-      final BalancerStrategy strategy = params.getBalancerStrategy();
-      if (availableSegments.contains(segment)) {
-        int assignedCount = assign(
-            params.getReplicationManager(),
-            tier,
-            totalReplicantsInCluster,
-            expectedReplicantsInTier,
-            totalReplicantsInTier,
-            strategy,
-            serverHolderList,
-            segment
+      final ServerHolder candidate = params.getBalancerStrategy().findNewSegmentHomeReplicator(segment, holders);
+      if (candidate == null) {
+        log.warn(
+            "Not enough [%s] servers or node capacity to assign primary segment[%s]! Expected Replicants[%d]",
+            tier, segment.getIdentifier(), targetReplicantsInTier
         );
-        stats.addToTieredStat(ASSIGNED_COUNT, tier, assignedCount);
-        totalReplicantsInCluster += assignedCount;
+      } else if (topCandidate == null || candidate.getServer().getPriority() > topCandidate.getServer().getPriority()) {
+        topCandidate = candidate;
       }
+    }
 
-      loadStatus.put(tier, expectedReplicantsInTier - loadedReplicantsInTier);
+    if (topCandidate != null) {
+      topCandidate.getPeon().loadSegment(segment, null);
     }
-    // Remove over-replication
-    stats.accumulate(drop(loadStatus, segment, params));
 
+    return topCandidate;
+  }
 
-    return stats;
+  private static void assignReplicas(
+      final Object2IntMap<String> targetReplicants,
+      final Object2IntMap<String> currentReplicants,
+      final DruidCoordinatorRuntimeParams params,
+      final DataSegment segment,
+      final CoordinatorStats stats,
+      @Nullable final String primaryTier
```

Comment:
Also maybe call it "tierToSkip", to make the intention more explicit.

Reply:
:+1: